In [1]:
from common import *

import gzip
import os
import shutil
import sqlite3
from urllib.parse import urlparse

import duckdb
import pandas as pd
import requests

from oaklib import get_adapter
from oaklib.datamodels.vocabulary import IS_A
from oaklib.utilities.lexical.lexical_indexer import create_lexical_index, save_lexical_index, load_lexical_index
from oaklib.interfaces.text_annotator_interface import TextAnnotatorInterface
from oaklib.implementations import AggregatorImplementation

from scipy.sparse import hstack
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report
from sklearn.model_selection import train_test_split

import yaml
import json


## todos
- [ ] cache goldterms.db
- [ ] cache goldterms.db **in notebooks/environmental_context_value_sets** with other cached files

In [2]:

selected_configuration = "soil_env_local_scale"

In [3]:
configurations_file = "voting_sheets_configurations.yaml"

In [4]:
# these are the variables that will be set by the configuration

gold_context_selectors = None
ncbi_context_selector = None
ncbi_package_selector = None
first_where = None
gold_first_where = None
semantic_anchor = None
previous_submission_schema_url = None
CONTEXT_ENUM = None
nmdc_package_selector = None
nmdc_context_selector = None
comparison_enum_column_name = None
output_file_name = None


In [5]:
# Load configurations from YAML file
with open(configurations_file, "r") as file:
    configurations = yaml.safe_load(file)["configurations"]


# Function to apply a configuration
def apply_configuration(config_name):
    config = configurations.get(config_name)
    if not config:
        raise ValueError(f"Configuration '{config_name}' not found.")

    # Update global variables
    globals().update(config)
    print(f"Configuration '{config_name}' applied.")


In [6]:
apply_configuration(selected_configuration)

# Now the global variables like `output_file_name`, `semantic_anchor`, etc., are set.

Configuration 'soil_env_local_scale' applied.


In [7]:
gold_first_where = first_where

In [8]:
# Initialize cache dictionaries for predict_from_normalized_env_packages
# todo how to move the definitions for function that use these globals? Or just use caching around the function?
ancestor_cache = {}
descendant_cache = {}

# Preliminary label, ancestor and descendant caching

In [9]:
# todo this on-demand NCBI curie extraction and annotation recapitulates work that is being added to
# https://portal.nersc.gov/project/m3408/biosamples_duckdb/
# via ???
#   although that doesn't detect auto-incremented curies from  spreadsheet dragging

# todo eventually, dig up a complete JSON gold biosample dump for non-hybrid biosample counts

# todo make it clearer whether biosamples or studies are being counted
#   count nmdc or gold STUDIES too?

# Additional Settings

In [10]:
# Approved prefixes (case-insensitive)
approved_prefixes = ['ENVO']

In [11]:
MIN_ANNOTATION_LEN = 3

In [12]:
NMDC_RUNTIME_BASE_URL = 'https://api.microbiomedata.org/nmdcschema/'
STUDY_SET_COLLECTION = 'study_set'
BIOSAMPLE_SET_COLLECTION = 'biosample_set'

In [13]:
envo_adapter_string = "sqlite:obo:envo"

In [14]:
po_adapter_string = "sqlite:obo:po"

In [15]:
env_package_override_file = 'mam-env-package-overrides.tsv'
override_column = 'mam_inferred_env_package'

In [16]:
ncbi_duckdb_url = 'https://portal.nersc.gov/project/m3408/biosamples_duckdb/ncbi_biosamples_2024-09-23.duckdb.gz'
# ncbi_duckdb_url = 'https://portal.nersc.gov/project/m3408/biosamples_duckdb/ncbi_biosamples.duckdb.gz'

In [17]:
gold_data_url = "https://gold.jgi.doe.gov/download?mode=site_excel"
gold_data_file_name = "goldData.xlsx"  # goldData.xlsx: Microsoft Excel 2007+
gold_csv_file_name = "gold_biosamples.csv"
BIOSAMPLES_SHEET = "Biosample"

In [18]:
goldterms_semsql_url = "https://s3.amazonaws.com/bbop-sqlite/goldterms.db.gz"

In [19]:
all_nmdc_biosamples_file = 'all_nmdc_biosamples.json'

# CURIe Constants

In [20]:
BIOME = 'ENVO:00000428'
TERRESTRIAL_BIOME = 'ENVO:00000446'
AQUATIC_BIOME = 'ENVO:00002030'
ABP = 'ENVO:01000813'
ENVIRONMENTAL_SYSTEM = 'ENVO:01000254'
ENVIRONMENTAL_MATERIAL = 'ENVO:00010483'

SOIL = 'ENVO:00001998'
LIQUID_WATER = 'ENVO:00002006'
WATER_ICE = 'ENVO:01000277'

HUMAN_CONSTRUCTION = 'ENVO:00000070'
BUILDING = 'ENVO:00000073'
BUILDING_PART = 'ENVO:01000420'

SNOW = 'ENVO:01000406'
MASS_OF_ENVIRONMENTAL_MATERIAL = 'ENVO:01001686'

PLANT_STRUCTURE = "PO:0009011" # does not include plant anatomical space or portion of plant substance (from plant anatomical entity)

MARINE_BIOME = 'ENVO:00000447'


# Settings-based Queries

In [21]:
# todo could this have been done with a OAK query, eliminating the need to explicitly download the file?

goldterms_envo_query = f"""
SELECT
	*
FROM
	statements s
WHERE
	predicate in ('{"', '".join(gold_context_selectors)}')"""

In [22]:
ncbi_biosamples_per_annotation_query = f"""
SELECT content, COUNT(1) AS count 
FROM attributes 
WHERE harmonized_name = '{ncbi_context_selector}' AND package_content like '%{ncbi_package_selector}'
GROUP BY content
ORDER BY COUNT(1) DESC
"""

In [23]:
ncbi_bioprojects_per_annotation_query = f"""
SELECT
	a.content,
	count(DISTINCT l.content) AS count
FROM
	main.ATTRIBUTES a
JOIN main.links l 
	ON
	a.id = l.id
WHERE
	l.target = 'bioproject'
	AND harmonized_name = 'env_local_scale'
	AND package_content like '%{ncbi_package_selector}'
GROUP BY
	a.content
ORDER BY
	count(DISTINCT l.content) DESC ;
"""

In [24]:
ncbi_X_per_annotation_query = ncbi_bioprojects_per_annotation_query

In [25]:
# and s1.subject = s1.stanza eliminates matches on blank node annotation rows (probably wouldn't change results but adds a little overhead)

extension_query = f"""
select
		s1.subject ,
		s2.predicate,
		COALESCE (s2."object",
	s2."value") as content
from
	statements s1
join statements s2 on 
	s1.subject = s2.subject
where
	{gold_first_where}
	and s1.predicate = 'rdfs:label'
	and s1.subject = s1.stanza
	and s2.predicate in ('mixs:env_broad', 'mixs:env_local', 'mixs:env_medium', 'mixs:mixs_extension', 'rdfs:label', 'mixs:other', 'mixs:anatomical_site', 'mixs:host_taxon') ;
"""


# Locally Defined Functions
_Currently using locally-defined cache dictionaries_

In [26]:
def predict_from_normalized_env_packages(df_raw, adapter):
    # Apply the function to the relevant columns

    df = df_raw.copy()

    print(df.shape)
    for column in ['env_broad_scale_id', 'env_local_scale_id', 'env_medium_id']:
        df[f'{column}_ancestors'] = df[column].apply(lambda x: get_hierarchy_terms(x, adapter)['ancestors'])
        df[f'{column}_descendants'] = df[column].apply(lambda x: get_hierarchy_terms(x, adapter)['descendants'])

    # Vectorize each set of terms separately
    broad_scale_ancestors = vectorize_terms(df, 'env_broad_scale_id_ancestors')
    broad_scale_descendants = vectorize_terms(df, 'env_broad_scale_id_descendants')

    local_scale_ancestors = vectorize_terms(df, 'env_local_scale_id_ancestors')
    local_scale_descendants = vectorize_terms(df, 'env_local_scale_id_descendants')

    medium_ancestors = vectorize_terms(df, 'env_medium_id_ancestors')
    medium_descendants = vectorize_terms(df, 'env_medium_id_descendants')

    # Combine all feature matrices
    X = hstack([
        broad_scale_ancestors,
        broad_scale_descendants,
        local_scale_ancestors,
        local_scale_descendants,
        medium_ancestors,
        medium_descendants
    ])

    # Filter the DataFrame to only include non-null rows for the target column
    df_filtered = df[df['normalized_env_package'].notnull() & (df['normalized_env_package'] != "")]

    # Extract the target variable
    y = df_filtered['normalized_env_package']

    # Ensure X corresponds to the filtered rows
    X_filtered = X[df_filtered.index]

    # Split the data into training and testing sets
    X_train, X_test, y_train, y_test = train_test_split(X_filtered, y, test_size=0.3, random_state=42)

    # Train a Random Forest Classifier
    clf = RandomForestClassifier(n_estimators=100, random_state=42)
    clf.fit(X_train, y_train)

    # Make predictions on the test set
    y_pred = clf.predict(X_test)

    # Evaluate the model
    print(classification_report(y_test, y_pred))

    # not determining confidence for each class nor saving any diagnostics any more

    return clf.predict(X)

In [27]:
def get_hierarchy_terms(my_curie: str, adapter) -> dict:
    """
    Extract ancestor and descendant terms from the ontology for a given CURIE,
    using caching to improve performance and filtering by 'is_a' relationships.

    Args:
        my_curie (str): CURIE identifier for the ontology term.
        adapter: Ontology adapter.

    Returns:
        dict: Dictionary containing lists of ancestor and descendant terms.
    """
    if my_curie not in ancestor_cache:
        try:
            ancestors = list(adapter.ancestors(my_curie, predicates=[IS_A]))
            ancestor_cache[my_curie] = [adapter.label(ancestor) for ancestor in ancestors if ancestor]
        except Exception as my_e:
            print(f"Error retrieving ancestors for {my_curie}: {my_e}")
            ancestor_cache[my_curie] = []

    if my_curie not in descendant_cache:
        try:
            descendants = list(adapter.descendants(my_curie, predicates=[IS_A]))
            descendant_cache[my_curie] = [adapter.label(descendant) for descendant in descendants if descendant]
        except Exception as my_e:
            print(f"Error retrieving descendants for {my_curie}: {my_e}")
            descendant_cache[my_curie] = []

    return {
        'ancestors': ancestor_cache[my_curie],
        'descendants': descendant_cache[my_curie],
    }

# Procedural Code Starts Here

In [28]:
# Determine the filenames and target directory for the NCBI DuckDB
ncbi_compressed_filename = urlparse(ncbi_duckdb_url).path.split('/')[-1]
ncbi_filename = os.path.splitext(ncbi_compressed_filename)[0]
ncbi_compressed_file_path = os.path.join(ncbi_compressed_filename)
ncbi_uncompressed_file_path = os.path.join(ncbi_filename)


In [29]:
if os.path.isfile(ncbi_uncompressed_file_path):
    print(f"{ncbi_uncompressed_file_path} is already present in the current working directory.")
else:
    if os.path.isfile(ncbi_compressed_file_path):
        print(f"{ncbi_compressed_file_path} is already present in the current working directory.")
    else:
        print(f"{ncbi_compressed_file_path} needs to be downloaded")
        ncbi_response = requests.get(ncbi_duckdb_url)
        with open(ncbi_compressed_file_path, "wb") as f:
            f.write(ncbi_response.content)
        # ~ 2 minutes @ 250 Mbps

    # Unzip the compressed file and save the extracted file in target directory
    print(f"{ncbi_compressed_file_path} needs to be unpacked")
    with gzip.open(ncbi_compressed_file_path, "rb") as f_in:
        with open(ncbi_uncompressed_file_path, "wb") as f_out:
            shutil.copyfileobj(f_in, f_out)

    # ~ 2 minutes

ncbi_biosamples_2024-09-23.duckdb is already present in the current working directory.


In [30]:
ncbi_conn = duckdb.connect(database=ncbi_uncompressed_file_path, read_only=True)

add po_adapter

should we be creating a list of adapters?

In [31]:
envo_adapter = get_adapter(envo_adapter_string)

In [32]:
po_adapter = get_adapter(po_adapter_string)

**Preliminary replacement for `get_hierarchy_terms` and ancestor/descendant lookup in main row construction loop**

In [33]:
reflexivity = False

# File paths
label_cache_file = "envo_label_cache.json"
ancestors_file = "envo_ancestors.json"
descendants_file = "envo_descendants.json"

# Initialize caches
envo_label_cache = {}
envo_ancs_cache = {}
envo_descs_cache = {}

envo_entities_generator = envo_adapter.entities()
envo_entities = list(envo_entities_generator)
envo_entities.sort()


# Function to generate and save the label cache
def generate_label_cache():
    print("Generating label cache...")

    label_cache = {}
    for ee in envo_entities:
        label_cache[ee] = envo_adapter.label(ee)
    return label_cache


# Check for label cache
if os.path.exists(label_cache_file):
    print(f"Loading label cache from {label_cache_file}...")
    with open(label_cache_file, "r") as f:
        envo_label_cache = json.load(f)
else:
    envo_label_cache = generate_label_cache()
    with open(label_cache_file, "w") as f:
        json.dump(envo_label_cache, f)
    print(f"Label cache saved to {label_cache_file}.")

# Check for ancestor cache
if os.path.exists(ancestors_file):
    print(f"Loading ancestors from {ancestors_file}...")
    with open(ancestors_file, "r") as f:
        envo_ancs_cache = json.load(f)
else:
    print("Ancestors file not found. Generating ancestor cache...")

    for ee in envo_entities:
        ancs_gen = envo_adapter.ancestors(ee, reflexive=reflexivity)
        envo_ancs_cache[ee] = [{a: envo_label_cache.get(a, None)} for a in ancs_gen]

    with open(ancestors_file, "w") as f:
        json.dump(envo_ancs_cache, f)
    print(f"Ancestors saved to {ancestors_file}.")

# Check for descendant cache
if os.path.exists(descendants_file):
    print(f"Loading descendants from {descendants_file}...")
    with open(descendants_file, "r") as f:
        envo_descs_cache = json.load(f)
else:
    # 5 minutes
    print("Descendants file not found. Generating descendant cache...")

    for ee in envo_entities:
        descs_gen = envo_adapter.descendants(ee, reflexive=reflexivity)
        envo_descs_cache[ee] = [{d: envo_label_cache.get(d, None)} for d in descs_gen]

    with open(descendants_file, "w") as f:
        json.dump(envo_descs_cache, f)
    print(f"Descendants saved to {descendants_file}.")


Loading label cache from envo_label_cache.json...
Loading ancestors from envo_ancestors.json...
Loading descendants from envo_descendants.json...


# Anchor aka bootstrapping classes

In [34]:
anchor_descendants = get_curie_descendants_label_dict(semantic_anchor, [IS_A], envo_adapter)

In [35]:
anchor_descendants_lod = curie_descendants_label_dict_to_lod(anchor_descendants)

In [36]:
anchor_descendants_frame = curie_descendants_label_lod_to_df(anchor_descendants_lod)

In [37]:
anchor_descendants_frame

curie                             label
0     ENVO:00001999                 marine water body
1     ENVO:01000188            tropical savanna biome
2     ENVO:00000487                  paternoster lake
3     ENVO:01000860  temperate marine upwelling biome
4     ENVO:01000199        mediterranean forest biome
...             ...                               ...
1731  ENVO:01000429                            burrow
1732  ENVO:01000431                      mixed forest
1733  ENVO:01000536                           factory
1734  ENVO:00000873                  freshwater biome
1735  ENVO:03500002                       public park

[1736 rows x 2 columns]

# Classes from the reference enumeration

In [38]:
sv = get_schemaview_from_source(previous_submission_schema_url)

In [39]:
# todo break out slow steps into its own cell

try:
    CONTEXT_ENUM_def = sv.get_enum(CONTEXT_ENUM)
    context_pvs_keys = list(CONTEXT_ENUM_def.permissible_values.keys())
except AttributeError as e:
    # Handle the AttributeError
    print(f"An AttributeError occurred: {e}")
    context_pvs_keys = []


In [40]:
print(context_pvs_keys)

['astronomical body part [ENVO:01000813]', '__coast [ENVO:01000687]', '__solid astronomical body part [ENVO:00000191]', '______landform [ENVO:01001886]', '______channel [ENVO:03000117]', '________tunnel [ENVO:00000068]', '______surface landform [ENVO:01001884]', '________desert [ENVO:01001357]', '________outcrop [ENVO:01000302]', '________boulder field [ENVO:00000537]', '________landfill [ENVO:00000533]', '________hummock [ENVO:00000516]', '________terrace [ENVO:00000508]', '________peninsula [ENVO:00000305]', '________shore [ENVO:00000304]', '__________lake shore [ENVO:00000382]', '________dry lake [ENVO:00000277]', '________karst [ENVO:00000175]', '________isthmus [ENVO:00000174]', '________badland [ENVO:00000127]', '________volcanic feature [ENVO:00000094]', '__________volcanic cone [ENVO:00000398]', '____________tuff cone [ENVO:01000664]', '________beach [ENVO:00000091]', '________plain [ENVO:00000086]', '________cave [ENVO:00000067]', '________spring [ENVO:00000027]', '______slope

In [41]:
initially_parsed_context_pvs = parse_hierarchically_underscored_strings(context_pvs_keys)

In [42]:
deduped_context_pvs = dedupe_underscoreless_pvs(initially_parsed_context_pvs)

In [43]:
pv_validation_results = validate_curie_label_list_dict(deduped_context_pvs, envo_adapter, print_flag=True)

In [44]:
pv_validation_results

{'problems': [],
 'valids': [{'curie': 'ENVO:01000813', 'label': 'astronomical body part'},
  {'curie': 'ENVO:01000687', 'label': 'coast'},
  {'curie': 'ENVO:00000191', 'label': 'solid astronomical body part'},
  {'curie': 'ENVO:01001886', 'label': 'landform'},
  {'curie': 'ENVO:03000117', 'label': 'channel'},
  {'curie': 'ENVO:00000068', 'label': 'tunnel'},
  {'curie': 'ENVO:01001884', 'label': 'surface landform'},
  {'curie': 'ENVO:01001357', 'label': 'desert'},
  {'curie': 'ENVO:01000302', 'label': 'outcrop'},
  {'curie': 'ENVO:00000537', 'label': 'boulder field'},
  {'curie': 'ENVO:00000533', 'label': 'landfill'},
  {'curie': 'ENVO:00000516', 'label': 'hummock'},
  {'curie': 'ENVO:00000508', 'label': 'terrace'},
  {'curie': 'ENVO:00000305', 'label': 'peninsula'},
  {'curie': 'ENVO:00000304', 'label': 'shore'},
  {'curie': 'ENVO:00000382', 'label': 'lake shore'},
  {'curie': 'ENVO:00000277', 'label': 'dry lake'},
  {'curie': 'ENVO:00000175', 'label': 'karst'},
  {'curie': 'ENVO:0000

# Get the CURIEs used in NMDC Biosample annotations

In [45]:
if os.path.isfile(all_nmdc_biosamples_file):
    print(
        f"{all_nmdc_biosamples_file} is present in the current working directory and will be read into all_nmdc_biosamples.")
    # with open(all_nmdc_biosamples_file, 'r') as file:
    #     all_nmdc_biosamples = yaml.full_load(file)
    # read as json
    with open(all_nmdc_biosamples_file, 'r') as f:
        all_nmdc_biosamples = json.load(f)

else:
    print(f"All NMDC Biosamples need to be fetched and saved to {all_nmdc_biosamples_file}")
    all_nmdc_biosamples = get_docs_from_nmdc_collection(NMDC_RUNTIME_BASE_URL,
                                                        BIOSAMPLE_SET_COLLECTION)
    # with open(all_nmdc_biosamples_file, 'w') as file:
    #     documents = yaml.dump(all_nmdc_biosamples, file)
    # save as json
    with open(all_nmdc_biosamples_file, 'w') as f:
        json.dump(all_nmdc_biosamples, f)

# this saves network traffic. could use JSON for faster performance. 
# 1 minute for network fetch and JSON write?!
# 1 minute for yaml read
# instantaneous for JSON read?

all_nmdc_biosamples.json is present in the current working directory and will be read into all_nmdc_biosamples.


## Prediction of env_package annotations 

In [46]:
# Specify the output file name
env_packages_file = "nmdc_biosample_asserted_normalized_and_inferred_env_package.tsv"

if os.path.exists(env_packages_file):
    # Load the DataFrame from the file if it exists
    print(f"Loading {env_packages_file} into nmdc_biosample_contexts_frame...")
    nmdc_biosample_contexts_frame = pd.read_csv(env_packages_file, sep='\t')
else:
    # File doesn't exist; generate the DataFrame
    print(f"{env_packages_file} not found. Predicting from asserted records and {env_package_override_file}...")

    # Load environment package overrides
    env_pacakge_overrides = tsv_to_dict_of_dicts(env_package_override_file, 'id')

    # Extract biosample contexts
    biosample_contexts_lod = biosamples_lod_context_extractor(
        all_nmdc_biosamples, envo_adapter,
        my_env_pacakge_overrides=env_pacakge_overrides
    )

    # Create the DataFrame
    nmdc_biosample_contexts_frame = pd.DataFrame(biosample_contexts_lod)

    # Print value counts for the 'normalized_env_package' column
    print("\n")
    print("Value counts for normalized_env_package column:")
    print(nmdc_biosample_contexts_frame['normalized_env_package'].value_counts(dropna=False))
    print("\n")

    # Generate package predictions
    package_predictions = predict_from_normalized_env_packages(nmdc_biosample_contexts_frame, envo_adapter)

    # Add predictions to the DataFrame
    nmdc_biosample_contexts_frame['predicted_env_package'] = package_predictions

    # Save the DataFrame to the file
    nmdc_biosample_contexts_frame.to_csv(env_packages_file, sep='\t', index=False)
    print(f"env_package predictions saved to {env_packages_file}")



Loading nmdc_biosample_asserted_normalized_and_inferred_env_package.tsv into nmdc_biosample_contexts_frame...


In [47]:
nmdc_biosample_contexts_frame['predicted_env_package'].value_counts()

predicted_env_package
soil                                               5916
water                                              1603
plant-associated                                    476
sediment                                            165
host-associated                                     156
miscellaneous natural or artificial environment     140
hydrocarbon resources-fluids_swabs                  114
Name: count, dtype: int64

## env-package prediction complete

To-do: save this and don't recreate it if it's available

Then get it reviewed by other NMDC stakeholders and inject it into MongoDB if approved

## Destructively filter `nmdc_biosample_contexts_frame` by `env_package` 

In [48]:
nmdc_biosample_contexts_frame.shape

(8570, 15)

In [49]:
nmdc_biosample_contexts_frame = nmdc_biosample_contexts_frame[
    nmdc_biosample_contexts_frame['predicted_env_package'] == nmdc_package_selector]

In [50]:
nmdc_biosample_contexts_frame.shape

(5916, 15)

# Long process of predicting OBO foundry CURIes from NCBI Biosamples

## Start by getting unique annotations? Pre-counted by Biosamples

Current task is to provide counts by "study" aka Bioproject in addition to Biosample counts or instead of Biosamples counts if necessary

In [51]:
ncbi_frame = ncbi_conn.execute(ncbi_X_per_annotation_query).fetchdf()


FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

In [52]:
ncbi_frame.insert(0, 'serial_number', range(1, len(ncbi_frame) + 1))

In [53]:
# includes values with counts of one... useful for discovering drag-down submissions?

## MIxS and NCBI guidelines imply environmental context slots are multivalued
and that the pipe `|` should be used as a delimiter

there's an envo_count value below that indicates how ofter other delimiters might be used

In [54]:
ncbi_frame['content_list'] = ncbi_frame['content'].str.split('|')

In [55]:
# todo is there any reason to not do this ?
ncbi_frame = ncbi_frame[ncbi_frame['content'].notna() & (ncbi_frame['content'] != '')]

In [56]:
ncbi_frame['content_count'] = ncbi_frame['content_list'].apply(len)

In [57]:
ncbi_frame.shape

(9820, 5)

In [58]:
ncbi_frame = ncbi_frame.explode('content_list').reset_index(drop=True)

In [59]:
ncbi_frame.shape

(10293, 5)

## splitting adds ~ 5% more rows
which might be important since were currently using a longest annotation strategy here

In [60]:
# how many content_list strings contain envo multiple times now?

In [61]:
ncbi_frame['envo_count'] = ncbi_frame['content_list'].str.lower().str.count("envo")

In [62]:
ncbi_frame['envo_count'].value_counts()

envo_count
0    8188
1    2094
2       9
3       1
4       1
Name: count, dtype: int64

## If my math is correct, about 0.1% of the annotations still contain multiple CURIes 
after splitting on pipes

There will also be annotations with multiple label-like strings that weren't split because they were not delimited on pipes
That might be a source of lost information since we are using a longest-match annotator here
I.e. there could be annotations with multiple hits worth keeping

## Parsing out CURIEs

this has a few limitations. The function only tries pre-specified prefixes (['ENVO'] by default) and only considers colons and underscores valid delimiters.

In [63]:
ncbi_frame[['extracted_label', 'extracted_curie']] = ncbi_frame['content_list'].apply(parse_curie_label)

In [64]:
parse_failures = ncbi_frame[
    (ncbi_frame['envo_count'] > 0) & (ncbi_frame['extracted_curie'].isna() | (ncbi_frame['extracted_curie'] == ''))]


## In what kinds of cases could no CURIe be parsed
despite the presence of "ENVO" in the content string?

In [65]:
parse_failures

serial_number                                      content  count  \
322              318                                    ENVO:Soil      3   
1398            1341                                  ENVO0000039      1   
1471            1412                                  ENVO0000071      1   
1472            1413                                  ENVO0000078      1   
1473            1414                                  ENVO0000229      1   
...              ...                                          ...    ...   
10246           9778                                  ENVO0000151      1   
10247           9779                                  ENVO0000208      1   
10248           9780                                  ENVO0000243      1   
10279           9809  forest soilENVO:00002261 Sampling point 105      1   
10280           9810  forest soilENVO:00002261 Sampling point 198      1   

                                      content_list  content_count  envo_count  \
322                                      ENVO:Soil              1           1   
1398                                   ENVO0000039              1           1   
1471                                   ENVO0000071              1           1   
1472                                   ENVO0000078              1           1   
1473                                   ENVO0000229              1           1   
...                                            ...            ...         ...   
10246                                  ENVO0000151              1           1   
10247                                  ENVO0000208              1           1   
10248                                  ENVO0000243              1           1   
10279  forest soilENVO:00002261 Sampling point 105              1           1   
10280  forest soilENVO:00002261 Sampling point 198              1           1   

                                   extracted_label extracted_curie  
322                                      ENVO Soil            None  
1398                                   ENVO0000039            None  
1471                                   ENVO0000071            None  
1472                                   ENVO0000078            None  
1473                                   ENVO0000229            None  
...                                            ...             ...  
10246                                  ENVO0000151            None  
10247                                  ENVO0000208            None  
10248                                  ENVO0000243            None  
10279  forest soilENVO 00002261 Sampling point 105            None  
10280  forest soilENVO 00002261 Sampling point 198            None  

[453 rows x 8 columns]

## Retrieve the labels for the parsed CURIes

In [66]:
ncbi_frame['real_label'] = ncbi_frame['extracted_curie'].apply(po_adapter.label)

## Apply oaklib annotation to the strings after CURIe removal
Actually the annotator can (sometimes?) detect colon-delimited CURIEs with lower case prefixes

This returns CURIes with evidence but not necessarily the label corresponding to the CURIe

In [67]:
# # Specify the lexical index file name
# aggregated_lexical_index_file = "aggregated_lexical_index.yaml"
# 
# # Check if the lexical index file exists
# if os.path.exists(aggregated_lexical_index_file):
#     print(f"Loading lexical index from {aggregated_lexical_index_file}...")
#     ix = load_lexical_index(aggregated_lexical_index_file)
# else:
#     print(f"{aggregated_lexical_index_file} not found. Creating lexical index from agg_adapter...")
#     ix = create_lexical_index(agg_adapter)
#     # Save the lexical index to a file
#     save_lexical_index(ix, aggregated_lexical_index_file)
#     print(f"Lexical index saved to {aggregated_lexical_index_file}")
# 
# # Initialize the TextAnnotatorInterface
# aggregated_text_annotator_interface = TextAnnotatorInterface()
# aggregated_text_annotator_interface.lexical_index = ix
# 
# # this cell only takes ~ 1 minute, but generates a lot of "ERRORS" and WARNINGS in a red font
# #   while lexically indexing the ontology

In [68]:
aggregated_text_annotator_interface = AggregatorImplementation(implementations=[envo_adapter, po_adapter])

In [69]:
# Apply the annotation function to each row in the 'label' column
ncbi_frame['longest_annotation_curie'] = ncbi_frame['extracted_label'].apply(
    lambda x: get_longest_annotation_curie(x, aggregated_text_annotator_interface, MIN_ANNOTATION_LEN))

# ~ 1 minute


ERROR:root:Skipping statements(subject=ENVO:00000112,predicate=oio:hasDbXref,object=<http://www.eionet.europa.eu/gemet/concept/8704>,value=None,datatype=None,language=None,); ValueError: <http://www.eionet.europa.eu/gemet/concept/8704> is not a valid URI or CURIE
ERROR:root:Skipping statements(subject=ENVO:00001996,predicate=oio:hasDbXref,object=<https://en.wikipedia.org/wiki/Acid_mine_drainage>,value=None,datatype=None,language=None,); ValueError: <https://en.wikipedia.org/wiki/Acid_mine_drainage> is not a valid URI or CURIE
ERROR:root:Skipping statements(subject=ENVO:01000225,predicate=oio:hasDbXref,object=<https://www.worldwildlife.org/biomes/tropical-and-subtropical-dry-broadleaf-forests>,value=None,datatype=None,language=None,); ValueError: <https://www.worldwildlife.org/biomes/tropical-and-subtropical-dry-broadleaf-forests> is not a valid URI or CURIE
ERROR:root:Skipping statements(subject=ENVO:01000227,predicate=oio:hasDbXref,object=<https://www.worldwildlife.org/biomes/tropical

## Add the labels for the CURIes identified though oaklib annotation of strings

In [70]:
ncbi_frame['longest_annotation_label'] = ncbi_frame['longest_annotation_curie'].apply(aggregated_text_annotator_interface.label)

In [71]:
ncbi_frame

serial_number                    content  count  \
0                  1             not applicable    269   
1                  2                    missing    205   
2                  3                       soil    203   
3                  4              not collected    170   
4                  5                     forest     81   
...              ...                        ...    ...   
10288           9816              SterileClosed      1   
10289           9817            cropping system      1   
10290           9818         near-rocky-outlier      1   
10291           9819  urban habitat, greenhouse      1   
10292           9820       Ironstone outcrop 01      1   

                    content_list  content_count  envo_count  \
0                 not applicable              1           0   
1                        missing              1           0   
2                           soil              1           0   
3                  not collected              1           0   
4                         forest              1           0   
...                          ...            ...         ...   
10288              SterileClosed              1           0   
10289            cropping system              1           0   
10290         near-rocky-outlier              1           0   
10291  urban habitat, greenhouse              1           0   
10292       Ironstone outcrop 01              1           0   

                 extracted_label extracted_curie real_label  \
0                 not applicable            None       None   
1                        missing            None       None   
2                           soil            None       None   
3                  not collected            None       None   
4                         forest            None       None   
...                          ...             ...        ...   
10288              SterileClosed            None       None   
10289            cropping system            None       None   
10290         near-rocky-outlier            None       None   
10291  urban habitat, greenhouse            None       None   
10292       Ironstone outcrop 01            None       None   

      longest_annotation_curie longest_annotation_label  
0                         None                     None  
1                         None                     None  
2                ENVO:00001998                     soil  
3                         None                     None  
4                ENVO:00000111            forested area  
...                        ...                      ...  
10288                     None                     None  
10289               RO:0002577                   system  
10290            ENVO:00001995                     rock  
10291            ENVO:03600087               greenhouse  
10292            ENVO:01000302                  outcrop  

[10293 rows x 11 columns]

## we now have a list of CURIes for each normalized annotation

This could be because the submitter provided a CURIe and a label that don't match
*One* case of this is dragging a CURIe down a column in a spreadsheet, expecting it to be copied,
but actually auto-incrementing it

Now attempt to find one best CURIe for each annotation... by now we have lost the ability to retain multiple legitimate
but improperly separated CURIes

In [72]:
# todo don't accept extracted curie if no real label?
# any kind of string similarity checking for label of annotated curie vs extracted label ?
# look for long stretches of curies?
# can we measure the beneficial impact of any of this? current crux: how to distribute counts

ncbi_frame['curie_list'] = ncbi_frame.apply(
    lambda my_row: list({my_row['extracted_curie'], my_row['longest_annotation_curie']} - {None}),
    axis=1
)

ncbi_frame['unique_curie_count'] = ncbi_frame['curie_list'].apply(len)

In [73]:
ncbi_frame['unique_curie_count'].value_counts()

unique_curie_count
1    5878
0    3859
2     556
Name: count, dtype: int64

In [74]:
double_curie_frame = ncbi_frame[ncbi_frame['unique_curie_count'] > 1]

In [75]:
double_curie_frame = double_curie_frame[['extracted_curie', 'longest_annotation_curie']]

In [76]:
double_curie_frame = double_curie_frame.drop_duplicates()

In [77]:
double_curie_frame[['extracted_prefix', 'extracted_local_id']] = double_curie_frame['extracted_curie'].str.split(':',
                                                                                                                 expand=True)

In [78]:
double_curie_frame['extracted_local_id_int'] = pd.to_numeric(double_curie_frame['extracted_local_id'],
                                                             errors='coerce').astype('Int64')

In [79]:
double_curie_frame

extracted_curie longest_annotation_curie extracted_prefix  \
175     ENVO:00000486             PATO:0000574             ENVO   
275      ENVO:3400002            ENVO:03400002             ENVO   
605     ENVO:01001430            ENVO:00000077             ENVO   
668     ENVO:00000114            ENVO:01000352             ENVO   
699     ENVO:01001357            ENVO:00000172             ENVO   
...               ...                      ...              ...   
10204   ENVO:00000414            ENVO:00000428             ENVO   
10252   ENVO:00000321            ENVO:00000428             ENVO   
10253   ENVO:00000377            ENVO:00000428             ENVO   
10277   ENVO:01001461            ENVO:01001803             ENVO   
10278   ENVO:01001469            ENVO:01001803             ENVO   

      extracted_local_id  extracted_local_id_int  
175             00000486                     486  
275              3400002                 3400002  
605             01001430                 1001430  
668             00000114                     114  
699             01001357                 1001357  
...                  ...                     ...  
10204           00000414                     414  
10252           00000321                     321  
10253           00000377                     377  
10277           01001461                 1001461  
10278           01001469                 1001469  

[439 rows x 5 columns]

In [80]:
# Ensure extracted_local_id_int is unique and sorted
unique_sorted_series = double_curie_frame['extracted_local_id_int'].dropna().drop_duplicates().sort_values()


In [81]:
# Find stretches
stretches_dict = find_consecutive_stretches_dict(unique_sorted_series)


In [82]:
# Convert the stretches dictionary into a DataFrame
stretches_df = stretches_dict_to_long_dataframe(stretches_dict)

`stretches_df` shows groups of extracted EnvoO ids (CURIes without prefix or padding zeros) that share a common CURIe by oaklib annotation of the textual part. This may not be the best or only way to address these spurious drag-stretch, auto-incremented CURIes

Ie 1001458 corresponds to ENVO:01001458, 'mist'

_although it theoretically could have been ENVO:1001458 since EnvO CURIes can have 7 or 8 digits_

In group 9, there are another ~ 50 sequential id values, all corresponding to environmental context annotations whose best oak-annotated class is ENVO:01001803, 'tropical forest'!

How much of an impact does this have? 

In [83]:
stretches_df

stretch_id    value
0             1        1
1             1        2
2             1        3
3             1        4
4             1        5
..          ...      ...
354           9  1001507
355           9  1001508
356           9  1001509
357           9  1001510
358           9  1001511

[359 rows x 2 columns]

In [84]:
# Perform the left merge
double_curie_frame = double_curie_frame.merge(
    stretches_df,
    left_on='extracted_local_id_int',
    right_on='value',
    how='left'
)

In [85]:
stretch_summary_df = summarize_stretch_groups(double_curie_frame)


For stretch 9, which included extracted CURIes from ENVO:01001458 to ENVO:01001511, the oaklib test annotation of 100% of the submitted environmental context annotations was ENVO:01001803, so we will keep that and disregard all of the CURIes from the stretch


In [86]:
stretch_summary_df

stretch_id most_common_longest_annotation_curie  fraction
0         1.0                        ENVO:01000703  0.683544
1         2.0                        ENVO:00000230  0.952381
2         3.0                        ENVO:00000428  0.993056
3         4.0                        ENVO:00000873  1.000000
4         5.0                        ENVO:00000078  1.000000
5         6.0                        ENVO:00002001  0.950000
6         7.0                        ENVO:01000431  0.333333
7         8.0                        ENVO:01000628  0.483871
8         9.0                        ENVO:01001803  1.000000

In [87]:
decisive_fraction_threshold = 0.9

In [88]:
decisive_stretch_summary_df = stretch_summary_df[stretch_summary_df['fraction'] >= decisive_fraction_threshold]

In [89]:
decisive_stretch_summary_df

stretch_id most_common_longest_annotation_curie  fraction
1         2.0                        ENVO:00000230  0.952381
2         3.0                        ENVO:00000428  0.993056
3         4.0                        ENVO:00000873  1.000000
4         5.0                        ENVO:00000078  1.000000
5         6.0                        ENVO:00002001  0.950000
8         9.0                        ENVO:01001803  1.000000

In [90]:
# Perform the left merge
double_curie_frame = double_curie_frame.merge(
    decisive_stretch_summary_df,
    left_on='stretch_id',
    right_on='stretch_id',
    how='left'
)

In [91]:
double_curie_frame

extracted_curie longest_annotation_curie extracted_prefix  \
0     ENVO:00000486             PATO:0000574             ENVO   
1      ENVO:3400002            ENVO:03400002             ENVO   
2     ENVO:01001430            ENVO:00000077             ENVO   
3     ENVO:00000114            ENVO:01000352             ENVO   
4     ENVO:01001357            ENVO:00000172             ENVO   
..              ...                      ...              ...   
434   ENVO:00000414            ENVO:00000428             ENVO   
435   ENVO:00000321            ENVO:00000428             ENVO   
436   ENVO:00000377            ENVO:00000428             ENVO   
437   ENVO:01001461            ENVO:01001803             ENVO   
438   ENVO:01001469            ENVO:01001803             ENVO   

    extracted_local_id  extracted_local_id_int  stretch_id      value  \
0             00000486                     486         NaN        NaN   
1              3400002                 3400002         NaN        NaN   
2             01001430                 1001430         NaN        NaN   
3             00000114                     114         NaN        NaN   
4             01001357                 1001357         NaN        NaN   
..                 ...                     ...         ...        ...   
434           00000414                     414         3.0      414.0   
435           00000321                     321         3.0      321.0   
436           00000377                     377         3.0      377.0   
437           01001461                 1001461         9.0  1001461.0   
438           01001469                 1001469         9.0  1001469.0   

    most_common_longest_annotation_curie  fraction  
0                                    NaN       NaN  
1                                    NaN       NaN  
2                                    NaN       NaN  
3                                    NaN       NaN  
4                                    NaN       NaN  
..                                   ...       ...  
434                        ENVO:00000428  0.993056  
435                        ENVO:00000428  0.993056  
436                        ENVO:00000428  0.993056  
437                        ENVO:01001803  1.000000  
438                        ENVO:01001803  1.000000  

[439 rows x 9 columns]

In [92]:
drag_evidence_frame = double_curie_frame[double_curie_frame['stretch_id'] >= 1]
drag_evidence_frame = drag_evidence_frame[['extracted_curie', 'longest_annotation_curie']]
drag_evidence_frame['drag_evidence'] = True

In [93]:
drag_evidence_frame

extracted_curie longest_annotation_curie  drag_evidence
6     ENVO:01000198            ENVO:01000431           True
9          ENVO:036            ENVO:01000703           True
10    ENVO:00000386            ENVO:00000428           True
11    ENVO:00000412            ENVO:00000428           True
12    ENVO:00000427            ENVO:00000428           True
..              ...                      ...            ...
434   ENVO:00000414            ENVO:00000428           True
435   ENVO:00000321            ENVO:00000428           True
436   ENVO:00000377            ENVO:00000428           True
437   ENVO:01001461            ENVO:01001803           True
438   ENVO:01001469            ENVO:01001803           True

[393 rows x 3 columns]

In [94]:
ncbi_frame = ncbi_frame.merge(
    drag_evidence_frame,
    left_on=['extracted_curie', 'longest_annotation_curie'],
    right_on=['extracted_curie', 'longest_annotation_curie'],
    how='left'
)

In [95]:
# Initialize dragless_curie_list with curie_list values
ncbi_frame["dragless_curie_list"] = ncbi_frame["curie_list"]

# Update dragless_curie_list based on the condition
for index, row in ncbi_frame.iterrows():
    if row["drag_evidence"] is True:
        if row["longest_annotation_curie"] is not None:
            ncbi_frame.at[index, "dragless_curie_list"] = [row["longest_annotation_curie"]]
        else:
            ncbi_frame.at[index, "dragless_curie_list"] = []

ncbi_frame['dragless_curie_count'] = ncbi_frame['dragless_curie_list'].apply(len)

In [96]:
ncbi_frame['unique_curie_count'].value_counts()

unique_curie_count
1    5878
0    3859
2     556
Name: count, dtype: int64

In [97]:
ncbi_frame['dragless_curie_count'].value_counts()

dragless_curie_count
1    6329
0    3859
2     105
Name: count, dtype: int64

## The extent of multiple detected CURIes has been reduced ~ 4.5 fold 
(for soil env_local_scale)

Isolate the submitter annotations for which there's clearly one best CURIe after removing the drag-stretches

In [98]:
ncbi_frame.shape

(10293, 16)

In [99]:
ncbi_frame_undisputed = ncbi_frame[ncbi_frame['dragless_curie_count'] <= 1]

In [100]:
ncbi_frame_undisputed.shape

(10188, 16)

In [101]:
ncbi_frame_disputed = ncbi_frame[ncbi_frame['dragless_curie_count'] > 1]

In [102]:
ncbi_frame_disputed.shape

(105, 16)

In [103]:
ncbi_frame_disputed = ncbi_frame_disputed.explode("dragless_curie_list", ignore_index=True)


In [104]:
ncbi_frame_disputed.shape

(210, 16)

In [105]:
ncbi_frame_disputed["dragless_curie_list"] = ncbi_frame_disputed["dragless_curie_list"].apply(lambda x: [x])

## Just include all of the remaining disputed CURIe assignments

In [106]:
# Combine the rows of ncbi_frame_undisputed and ncbi_frame_disputed into a new DataFrame
ncbi_disputes_exploded_frame = pd.concat([ncbi_frame_undisputed, ncbi_frame_disputed], ignore_index=True)


In [107]:
ncbi_disputes_exploded_frame.shape

(10398, 16)

In [108]:
ncbi_disputes_exploded_frame

serial_number                                            content  \
0                  1                                     not applicable   
1                  2                                            missing   
2                  3                                               soil   
3                  4                                      not collected   
4                  5                                             forest   
...              ...                                                ...   
10393           9710                               soil [ENVO:01001044]   
10394           9767  irrigation system [ENVO:00000029]|fertilizatio...   
10395           9767  irrigation system [ENVO:00000029]|fertilizatio...   
10396           9813  irrigation system [ENVO:00000029]|fertilizatio...   
10397           9813  irrigation system [ENVO:00000029]|fertilizatio...   

       count                           content_list  content_count  \
0        269                         not applicable              1   
1        205                                missing              1   
2        203                                   soil              1   
3        170                          not collected              1   
4         81                                 forest              1   
...      ...                                    ...            ...   
10393      1                   soil [ENVO:01001044]              1   
10394      1  fertilization process [ENVO:02500001]              3   
10395      1  fertilization process [ENVO:02500001]              3   
10396      1  fertilization process [ENVO:02500001]              3   
10397      1  fertilization process [ENVO:02500001]              3   

       envo_count        extracted_label extracted_curie real_label  \
0               0         not applicable            None       None   
1               0                missing            None       None   
2               0                   soil            None       None   
3               0          not collected            None       None   
4               0                 forest            None       None   
...           ...                    ...             ...        ...   
10393           1                   soil   ENVO:01001044       None   
10394           1  fertilization process   ENVO:02500001       None   
10395           1  fertilization process   ENVO:02500001       None   
10396           1  fertilization process   ENVO:02500001       None   
10397           1  fertilization process   ENVO:02500001       None   

      longest_annotation_curie longest_annotation_label  \
0                         None                     None   
1                         None                     None   
2                ENVO:00001998                     soil   
3                         None                     None   
4                ENVO:00000111            forested area   
...                        ...                      ...   
10393            ENVO:00001998                     soil   
10394              BFO:0000015                  process   
10395              BFO:0000015                  process   
10396              BFO:0000015                  process   
10397              BFO:0000015                  process   

                           curie_list  unique_curie_count drag_evidence  \
0                                  []                   0           NaN   
1                                  []                   0           NaN   
2                     [ENVO:00001998]                   1           NaN   
3                                  []                   0           NaN   
4                     [ENVO:00000111]                   1           NaN   
...                               ...                 ...           ...   
10393  [ENVO:01001044, ENVO:00001998]                   2           NaN   
10394    [BFO:0000015, ENVO:02500001]                   2           NaN   
10395    [BFO:0000015, ENVO:

In [109]:
ncbi_disputes_exploded_frame['post_explode_curie_count'] = ncbi_disputes_exploded_frame['dragless_curie_list'].apply(
    len)

In [110]:
ncbi_disputes_exploded_frame['post_explode_curie_count'].value_counts()

post_explode_curie_count
1    6539
0    3859
Name: count, dtype: int64

In [111]:
# Set 'post_explode_curie' to the 0th item in 'dragless_curie_list'
ncbi_disputes_exploded_frame["post_explode_curie"] = ncbi_disputes_exploded_frame["dragless_curie_list"].apply(
    lambda x: x[0] if isinstance(x, list) and len(x) > 0 else None
)

In [112]:

ncbi_biosample_scoped_counts = (
    ncbi_disputes_exploded_frame.groupby("post_explode_curie")["count"].sum().reset_index()
)

# todo parameterize the column name based on what was really counted, generally biosamples or bioprojects
ncbi_biosample_scoped_counts.columns = ['curie', 'ncbi_bioprojects_count']

This is currently a count of Biosamples for which the indicated CURIes can be extracted or inferred by oaklib annotation, after removal of drag-stretch, auto-incremented CURIes

In [113]:
ncbi_biosample_scoped_counts

curie  ncbi_bioprojects_count
0        BFO:0000015                      62
1        BFO:0000029                      63
2        BFO:0000030                       1
3        CHEBI:15377                      51
4        CHEBI:15379                       1
...              ...                     ...
1099  UBERON:0000922                       1
1100  UBERON:0001456                       3
1101  UBERON:0001474                       2
1102  UBERON:0001988                       5
1103  UBERON:0002416                      98

[1104 rows x 2 columns]

## GOLDTERMS mappings only approach

we're currently including
- mappings only

and have retired the previous use of
- mappings in hybrid with biosample counts

And we're casting a wide net, especially for the hybrid approach
- searching for 'soil', 'sediment' etc. in GOLDTERMS labels without anchoring them like 'Environmental > Aquatic > Sediment'
- retrieving the CURIes for env_broad_scale, env_local_scale and env_medium for all voting sheets, and trusting orthogonal filtering to remove the inappropriate CURIes

Should we now add (or switch to) direct biosample counts of GOLD "envo" annotations?

Efficient retrieval of  all GOLD data in a given scope isn't easy

In [114]:
# Determine the filenames and target directory
goldterms_compressed_filename = urlparse(goldterms_semsql_url).path.split('/')[-1]
goldterms_filename = os.path.splitext(goldterms_compressed_filename)[0]
target_dir = os.path.join("..", "..")  # Two levels up

# Print to confirm the filenames
print(goldterms_filename)

goldterms.db


In [115]:
# Fetch the contents from the URL and save compressed file in target directory
goldterms_response = requests.get(goldterms_semsql_url)
goldterms_compressed_file_path = os.path.join(target_dir, goldterms_compressed_filename)
with open(goldterms_compressed_file_path, "wb") as f:
    f.write(goldterms_response.content)

# ~ 1 second

In [116]:
# Unzip the compressed file and save the extracted file in target directory
goldterms_uncompressed_file_path = os.path.join(target_dir, goldterms_filename)
with gzip.open(goldterms_compressed_file_path, "rb") as f_in:
    with open(goldterms_uncompressed_file_path, "wb") as f_out:
        shutil.copyfileobj(f_in, f_out)

# ~ 1 second

In [117]:
## that's all fast. don't bother caching

In [118]:
goldterms_conn = sqlite3.connect(goldterms_uncompressed_file_path)

In [119]:
goldterms_result = pd.read_sql_query(extension_query, goldterms_conn)

In [120]:
goldterms_result

subject            predicate  \
0                       <GOLDVOCAB:Paddy-field/soil>           rdfs:label   
1    GOLDTERMS:Environmental-Terrestrial-Soil-Arable       mixs:env_broad   
2    GOLDTERMS:Environmental-Terrestrial-Soil-Arable       mixs:env_local   
3    GOLDTERMS:Environmental-Terrestrial-Soil-Arable      mixs:env_medium   
4    GOLDTERMS:Environmental-Terrestrial-Soil-Arable  mixs:mixs_extension   
..                                               ...                  ...   
458                          GOLDVOCAB:Riparian-soil           rdfs:label   
459                                   GOLDVOCAB:Soil           rdfs:label   
460                 GOLDVOCAB:Soil-_28non-planted_29           rdfs:label   
461                             GOLDVOCAB:Soil-crust           rdfs:label   
462                         GOLDVOCAB:Soil-microcosm           rdfs:label   

                content  
0      Paddy field/soil  
1         ENVO:00000446  
2         ENVO:01001177  
3         ENVO:00005742  
4             mixs:Soil  
..                  ...  
458       Riparian soil  
459                Soil  
460  Soil (non-planted)  
461          Soil crust  
462      Soil microcosm  

[463 rows x 3 columns]

In [121]:
# # todo: save this kind of content before subsetting on an environment
# #   the subsetting is currently baked into the query
# 
# # see also goldterms_queries.ipynb in MAM's Collab
# goldterms_result.to_csv("goldterms_single_environment_mappings_long.tsv", sep="\t", index=False)

In [122]:
goldterms_only_curies = goldterms_result.loc[goldterms_result['predicate'].isin(gold_context_selectors), 'content']


In [123]:
goldterms_only_curies = goldterms_only_curies.unique().tolist()

In [124]:
# goldterms_only_curies

# Make lists of CURIEs
which will determine
- the rows in the table
- the boolean filter columns

this next block got a lot slower when I switched from the envo adapter to the aggregated adapter

In [125]:
anchor_curies = list(anchor_descendants_frame['curie'])


In [126]:
legacy_pv_curies = [i['curie'] for i in pv_validation_results['valids']]

In [127]:
obsoletes_curies_envo = set(envo_adapter.obsoletes())
obsoletes_curies_po = set(po_adapter.obsoletes())
obsoletes_curies = list(obsoletes_curies_envo.union(obsoletes_curies_po))

In [128]:
abp_curies = list(envo_adapter.descendants(ABP, predicates=[IS_A]))
aquatic_biome_curies = list(envo_adapter.descendants(AQUATIC_BIOME, predicates=[IS_A]))
biome_curies = list(envo_adapter.descendants(BIOME, predicates=[IS_A]))  # 
building_curies = list(envo_adapter.descendants(BUILDING, predicates=[IS_A]))  #
building_part_curies = list(envo_adapter.descendants(BUILDING_PART, predicates=[IS_A]))  #
env_mat_curies = list(envo_adapter.descendants(ENVIRONMENTAL_MATERIAL, predicates=[IS_A]))
env_sys_curies = list(envo_adapter.descendants(ENVIRONMENTAL_SYSTEM, predicates=[IS_A]))
env_sys_curies = list(envo_adapter.descendants(ENVIRONMENTAL_SYSTEM, predicates=[IS_A]))
human_construction_curies = list(envo_adapter.descendants(HUMAN_CONSTRUCTION, predicates=[IS_A]))  #
liquid_water_curies = list(envo_adapter.descendants(LIQUID_WATER, predicates=[IS_A]))  # 
marine_biome_curies = list(envo_adapter.descendants(MARINE_BIOME, predicates=[IS_A]))
mass_of_environmental_material_curies = list(envo_adapter.descendants(MASS_OF_ENVIRONMENTAL_MATERIAL, predicates=[IS_A]))  #
plant_structure_curies = list(po_adapter.descendants(PLANT_STRUCTURE, predicates=[IS_A]))
snow_curies = list(envo_adapter.descendants(SNOW, predicates=[IS_A]))  #
soil_curies = list(envo_adapter.descendants(SOIL, predicates=[IS_A]))  #
terrestrial_biome_curies = list(envo_adapter.descendants(TERRESTRIAL_BIOME, predicates=[IS_A]))
water_ice_curies = list(envo_adapter.descendants(WATER_ICE, predicates=[IS_A]))  #


In [129]:
# Specify an output file name
nlcd2011_class_iris_file = "nlcd2011_class_iris.txt"

nlcd_subset_textual_representation = "nlcd2011"

in_subset_curie = "oio:inSubset"

# Initialize nlcd_classes
nlcd_classes = []

if os.path.exists(nlcd2011_class_iris_file):
    # Load the list from the file if it exists
    print(f"Loading {nlcd_subset_textual_representation} classes from {nlcd2011_class_iris_file}...")
    with open(nlcd2011_class_iris_file, "r") as file:
        nlcd_classes = [line.strip() for line in file.readlines()]
else:
    # File doesn't exist; generate the list using the loop
    print(
        f"{nlcd2011_class_iris_file} not found. Identifying classes in {nlcd_subset_textual_representation} subset (~2 minutes)...")

    # Retrieve all classes
    entities = envo_adapter.entities()

    # super slow 2 minutes
    # but retrieving classes by named subset seems to crash on EnvO with its textual subsets?
    for entity in entities:
        term_metadata = envo_adapter.entity_metadata_map(entity)
        if in_subset_curie in term_metadata:
            subsets = term_metadata[in_subset_curie]
            if nlcd_subset_textual_representation in subsets:
                nlcd_classes.append(entity)

    # Save the generated list to the file
    with open(nlcd2011_class_iris_file, "w") as file:
        for string in nlcd_classes:
            file.write(string + "\n")

    print(f"List saved to {nlcd2011_class_iris_file}")

# At this point, nlcd_classes contains the desired list
print(f"Total {nlcd_subset_textual_representation} classes loaded: {len(nlcd_classes)}")


Loading nlcd2011 classes from nlcd2011_class_iris.txt...
Total nlcd2011 classes loaded: 22


## Bootstrap the rows

In [130]:
include_in_rows = set()

In [131]:
include_in_rows.update(anchor_curies)
include_in_rows.update(goldterms_only_curies)
include_in_rows.update(legacy_pv_curies)
include_in_rows.update(ncbi_frame['extracted_curie'])
include_in_rows.update(ncbi_frame['longest_annotation_curie'])
include_in_rows.update(nlcd_classes)
include_in_rows.update(nmdc_biosample_contexts_frame[nmdc_context_selector])

In [132]:
rows_lod = []

# Voting sheet rows and boolean columns

In [133]:
for curie in include_in_rows:
    if curie is None:
        continue

    # ONCE AGAIN, assuming that EnvO is the only ontology we'll check against
    current_ancestors = [
        list(ancestor.keys())[0] for ancestor in envo_ancs_cache.get(curie, [])
    ]
    ancestors_in_enum_count = len(set(current_ancestors) & set(legacy_pv_curies))

    current_descendants = [
        list(descendant.keys())[0] for descendant in envo_descs_cache.get(curie, [])
    ]
    descendants_in_enum_count = len(set(current_descendants) & set(legacy_pv_curies))

    row = {
        'curie': curie,
        'claimed_prefix': curie.split(':')[0],
        'label': aggregated_text_annotator_interface.label(curie),
        'envo_native': False,
        'obsolete': False,
        comparison_enum_column_name: False,
        'ancestors_in_enum_count': ancestors_in_enum_count,
        'descendants_in_enum_count': descendants_in_enum_count,
        'nlcd_class': False,
        'abp': False,
        'env_sys': False,
        'biome': False,
        'terrestrial_biome': False,
        'aquatic_biome': False,
        'marine_biome': False,
        'env_mat': False,
        'soil': False,
        'liquid water': False,
        'water ice': False,
        'snow': False,
        'mass_of_environmental_material': False,
        'human_construction': False,
        'building': False,
        'building_part': False,
        'plant_structure': False,
        'goldterms_mappings': False,
    }

    if curie in abp_curies: row['abp'] = True
    if curie in aquatic_biome_curies: row['aquatic_biome'] = True
    if curie in biome_curies: row['biome'] = True
    if curie in building_curies: row['building'] = True
    if curie in building_part_curies: row['building_part'] = True
    if curie in env_mat_curies: row['env_mat'] = True
    if curie in env_sys_curies: row['env_sys'] = True
    if curie in goldterms_only_curies: row['goldterms_mappings'] = True
    if curie in human_construction_curies: row['human_construction'] = True
    if curie in legacy_pv_curies: row[comparison_enum_column_name] = True
    if curie in liquid_water_curies: row['liquid water'] = True
    if curie in marine_biome_curies: row['marine_biome'] = True
    if curie in mass_of_environmental_material_curies: row['mass_of_environmental_material'] = True
    if curie in nlcd_classes: row['nlcd_class'] = True
    if curie in obsoletes_curies: row['obsolete'] = True
    if curie in plant_structure_curies: row['plant_structure'] = True
    if curie in snow_curies: row['snow'] = True
    if curie in soil_curies: row['soil'] = True
    if curie in terrestrial_biome_curies: row['terrestrial_biome'] = True
    if curie in water_ice_curies: row['water ice'] = True

    try:
        prefix, local_id = curie.split(':')
        if prefix and prefix == 'ENVO' and row['label'] is not None:
            row['envo_native'] = True
    except Exception as e:
        # Print the exception message
        print(f"An error occurred: {e} trying to split {curie}")

    rows_lod.append(row)

# 2 minutes


# ^ Voting sheet rows and boolean columns

In [134]:
rows_frame = pd.DataFrame(rows_lod)

In [135]:
rows_frame

curie claimed_prefix                    label  envo_native  \
0     ENVO:03501301           ENVO                baseboard         True   
1     ENVO:00003042           ENVO                  piggery         True   
2     ENVO:01001335           ENVO               anchor ice         True   
3        ENVO:01014           ENVO                     None        False   
4     ENVO:01001494           ENVO                 moonmoon         True   
...             ...            ...                      ...          ...   
2516  ENVO:01000061           ENVO  marine wind mixed layer         True   
2517  ENVO:01000745           ENVO   area of perennial snow         True   
2518   ENVO:0000546           ENVO                     None        False   
2519  ENVO:00000037           ENVO                    ditch         True   
2520  ENVO:01000312           ENVO    culturing environment         True   

      obsolete  EnvLocalScaleSoilEnum_10_7  ancestors_in_enum_count  \
0        False                       False                        1   
1        False                       False                        1   
2        False                       False                        4   
3        False                       False                        0   
4        False                       False                        0   
...        ...                         ...                      ...   
2516     False                       False                        1   
2517     False                       False                        1   
2518     False                       False                        0   
2519     False                        True                        8   
2520     False                       False                        0   

      descendants_in_enum_count  nlcd_class    abp  ...   soil  liquid water  \
0                             0       False   True  ...  False         False   
1                             0       False   True  ...  False         False   
2                             0       False   True  ...  False         False   
3                             0       False  False  ...  False         False   
4                             0       False  False  ...  False         False   
...                         ...         ...    ...  ...    ...           ...   
2516                          0       False   True  ...  False         False   
2517                          0        True   True  ...  False         False   
2518                          0       False  False  ...  False         False   
2519                          0       False   True  ...  False         False   
2520                          0       False  False  ...  False         False   

      water ice   snow  mass_of_environmental_material  human_construction  \
0         False  False                           False                True   
1         False  False                           False                True   
2         False  False                            True               False   
3         False  False                           False               False   
4         False  False                           False               False   
...         ...    ...                             ...                 ...   
2516      False  False                           False               False   
2517      False  False                           False               False   
2518      False  False                           False               False   
2519      False  False                           False                True   
2520      False  False                           False               False   

      building  building_part  plant_structure  goldterms_mappings  
0        False           True            False               False  
1         True          False            False               False  
2        False          False            False               False  
3        False          False            False               False  
4    

# Merge in NMDC counts

In [136]:
nmdc_biosample_scoped_counts = nmdc_biosample_contexts_frame[nmdc_context_selector].value_counts().reset_index()
# todo parameterize the column name based on what what really counter (generally biosamples or studies)
nmdc_biosample_scoped_counts.columns = ['curie', 'nmdc_biosamples_count']


In [137]:
nmdc_biosample_scoped_counts

curie  nmdc_biosamples_count
0   ENVO:01000843                   1268
1   ENVO:01000816                   1067
2   ENVO:01000888                    777
3   ENVO:01000869                    614
4   ENVO:01000861                    295
5   ENVO:01000893                    245
6   ENVO:01000892                    207
7   ENVO:01000855                    199
8   ENVO:01000891                    169
9   ENVO:01000887                    123
10  ENVO:00005801                    118
11  ENVO:00000011                    113
12  ENVO:00000044                    107
13  ENVO:01001803                    105
14  ENVO:01000219                     83
15  ENVO:00000376                     60
16  ENVO:00000292                     53
17  ENVO:01000894                     45
18  ENVO:00000516                     38
19  ENVO:03600018                     28
20  ENVO:01000355                     18
21  ENVO:00005784                     17
22  ENVO:03600011                     14
23  ENVO:00002000                     14
24  ENVO:01001785                     14
25  ENVO:01000174                     13
26  ENVO:00000114                     12
27  ENVO:00000100                     12
28  ENVO:01001370                     12
29  ENVO:00002204                     10
30  ENVO:00000291                     10
31  ENVO:00000120                      6
32  ENVO:01000185                      6
33  ENVO:00000119                      4
34  ENVO:00000086                      4
35  ENVO:00000444                      4
36  ENVO:00000182                      4
37  ENVO:00000305                      4
38  ENVO:00000083                      4
39  ENVO:00000094                      4
40  ENVO:01000247                      3
41  ENVO:00000161                      3
42  ENVO:01000211                      3
43  ENVO:00000128                      2
44  ENVO:01000221                      2
45  ENVO:00000170                      1
46  ENVO:01000215                      1
47  ENVO:00000548                      1

In [138]:
# Perform the left merge
rows_frame = rows_frame.merge(
    nmdc_biosample_scoped_counts,
    left_on='curie',
    right_on='curie',
    how='left'
)

In [139]:
rows_frame

curie claimed_prefix                    label  envo_native  \
0     ENVO:03501301           ENVO                baseboard         True   
1     ENVO:00003042           ENVO                  piggery         True   
2     ENVO:01001335           ENVO               anchor ice         True   
3        ENVO:01014           ENVO                     None        False   
4     ENVO:01001494           ENVO                 moonmoon         True   
...             ...            ...                      ...          ...   
2516  ENVO:01000061           ENVO  marine wind mixed layer         True   
2517  ENVO:01000745           ENVO   area of perennial snow         True   
2518   ENVO:0000546           ENVO                     None        False   
2519  ENVO:00000037           ENVO                    ditch         True   
2520  ENVO:01000312           ENVO    culturing environment         True   

      obsolete  EnvLocalScaleSoilEnum_10_7  ancestors_in_enum_count  \
0        False                       False                        1   
1        False                       False                        1   
2        False                       False                        4   
3        False                       False                        0   
4        False                       False                        0   
...        ...                         ...                      ...   
2516     False                       False                        1   
2517     False                       False                        1   
2518     False                       False                        0   
2519     False                        True                        8   
2520     False                       False                        0   

      descendants_in_enum_count  nlcd_class    abp  ...  liquid water  \
0                             0       False   True  ...         False   
1                             0       False   True  ...         False   
2                             0       False   True  ...         False   
3                             0       False  False  ...         False   
4                             0       False  False  ...         False   
...                         ...         ...    ...  ...           ...   
2516                          0       False   True  ...         False   
2517                          0        True   True  ...         False   
2518                          0       False  False  ...         False   
2519                          0       False   True  ...         False   
2520                          0       False  False  ...         False   

      water ice   snow  mass_of_environmental_material  human_construction  \
0         False  False                           False                True   
1         False  False                           False                True   
2         False  False                            True               False   
3         False  False                           False               False   
4         False  False                           False               False   
...         ...    ...                             ...                 ...   
2516      False  False                           False               False   
2517      False  False                           False               False   
2518      False  False                           False               False   
2519      False  False                           False                True   
2520      False  False                           False               False   

      building  building_part  plant_structure  goldterms_mappings  \
0        False           True            False               False   
1         True          False            False               False   
2        False          False            False               False   
3        False          False            False               False   
4        False          False            False               False   
...        ...

In [140]:
# gold and ncbi counts are slightly trickier
# for gold: including mappings only, mappings in hybrid with biosample counts. 
#    Switch to direct biosample counts of GOLD "envo" annotations?
# ncbi: we have extracted curies and annotated curies

## Merge in NCBI counts

In [141]:
# Perform the left merge
rows_frame = rows_frame.merge(
    ncbi_biosample_scoped_counts,
    left_on='curie',
    right_on='curie',
    how='left'
)

In [142]:
# 990 rows in https://docs.google.com/spreadsheets/d/12WH3eduBq2qSTy9zVF3n7fyajn6ssLZL/edit?gid=546570706#gid=546570706

In [143]:
rows_frame.to_csv(output_file_name, sep="\t", index=False)

In [144]:
ncbi_conn.close()